### ISSUE TITLE
Method overload doesn't resolve correctly resolve when using union type

### ISSUE THREAD

**Issue by mhofman** [microsoft/TypeScript#40181]
**TypeScript Version:**  4.0.2

**Search Terms:** overload union type callback

**Expected behavior:** No Type error

**Actual behavior:** `Argument of type '"change"' is not assignable to parameter of type '"focus"'`

**Related Issues:** Possibly #39692

**Code**
```ts
interface ChangeEvent {
  type: 'foo' | 'bar';
  empty: boolean;
}

interface FooChangeEvent extends ChangeEvent {
  type: 'foo';
  fooInfo: string;
}

interface BarChangeEvent extends ChangeEvent {
  type: 'bar';
  barInfo: number;
}

interface Base {
  clear(): Base;
}

interface Foo extends Base {
  on(eventType: 'focus', handler: () => void): Foo;
  on(eventType: 'change', handler: (event: FooChangeEvent) => void): Foo;
}

interface Bar extends Base {
  on(eventType: 'focus', handler: () => void): Bar;
  on(eventType: 'change', handler: (event: BarChangeEvent) => void): Bar;
}

let foo: Foo = ({} as Foo);
foo.on('focus', () => {});
foo.on('change', () => {});

let bar: Bar = ({} as Bar);
bar.on('focus', () => {});
bar.on('change', () => {});

function handleEvents(obj: Foo | Bar) {
  obj.on('focus', () => {});
  obj.on('change', () => {});
}
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
let foo = {};
foo.on('focus', () => { });
foo.on('change', () => { });
let bar = {};
bar.on('focus', () => { });
bar.on('change', () => { });
function handleEvents(obj) {
    obj.on('focus', () => { });
    obj.on('change', () => { });
}

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIAs4gOYQKIBuE4yA3gFDLJgCeADhAFzIDkMA9uy8gD6sBGcKCwDclZBAC2dWs36cANhCxiAvuXKhIsRCgBinDFlyFiYCQA9IIACYBnNJhz4iJClVoNmbTqPEd2AEkQDmY7MChQbDUNLWh4JGQAISEjZ1MSCCtie0djFzMycU8mASE-KkEoYNDkEABXSX5oGM1weN1kuDsUd2QEJSEACgBKZhSe1ridRIN2S2tcid7xdhAhiFcwABV6Up8EersWABpkJxslKGZR5ABeAD5kAnZgGzHkObEqNY2t3a8rAQTlwp3OWEu0BumzMzDmaRMWxG9yeLzeHy+5HUbW0CRQKSgCxyDmWRR+6xh4AB+w4h2OZwuVxuyMez1e73GQm+yF+lJ2e28wPyYMZUOQf1hXSgCIK4BZqPZHwJrSU5gCcM493FpFUyG6n04IzEAQAdL8DkcwbdWTqjeRTeahc4rfKyKo7eRVcgqpzCXdtbr9QS7VUzesLfTxa7bWJQ46QRAXSi3R6YPUQAgwMA1uDbEoMmA7EN2PwAFYa+Z8YNknllsNDCNJm3u7kl0v1lhO0Fna1PGNYoA)
      

## Details

[Removing the overload](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIAs4gOYQKIBuE4yA3gFDLJgCeADhAFzIDkMA9uy8gD6sBGcKCwDclZBAC2dWs36cANhCxiAvuXKhIsRCgBinDFlyFiYCQA9IIACYBnNJhz4iJClVoNmbTqPEd2AEkQDmY7MChQbDUNLWh4JGQAISEjZ1MSCCtie0djFzMycU8mASE-KkEoYNDkEABXSX5oGM1weN1kuDsUd2QEJSEACgBKZhSe1ridRIN2S2tcid7xAHpV9hAhiFcwABV6Up8EersWABpkJxslKGZR5ABeAD5kAnZgGzHkObEqTe2uwOXlYCCcuAuVywN2g9x2ZmYczSJl2Iyer3en2+v3I6ja2gSKBSUAWOQcyyKVHWAPh4GBRw4JzOl2ut3uaJebw+X3GQj+yBpQMO3jB+UhrNhyEBCK6UGRBXAHIx3O+xNaSnMAURnCeUtIqmQ3R+nBGYnWAQAdADjqdIQ9OfrTeRLdbRc47UqyKoneQNcgqrySY89QajcSnesqlatjbmVLPY6xFHXeCIB70V6fTB6iAEGBgJsobYlBkwHYhux+AArbXzPjhynIanV6NDWPph3e-mVqutlhuiGXe2vRO4oA) (commenting out the `focus` eventType in the example), work fine, so TS is able to resolve the union type for the handler.

[Adding an `event` argument as a union `FooChangeEvent | BarChangeEvent` to the change's handler](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIAs4gOYQKIBuE4yA3gFDLJgCeADhAFzIDkMA9uy8gD6sBGcKCwDclZBAC2dWs36cANhCxiAvuXKhIsRCgBinDFlyFiYCQA9IIACYBnNJhz4iJClVoNmbTqPEd2AEkQDmY7MChQbDUNLWh4JGQAISEjZ1MSCCtie0djFzMycU8mASE-KkEoYNDkEABXSX5oGM1weN1kuDsUd2QEJSEACgBKZhSe1ridRIN2S2tcid7xdhAhiFcwABV6Up8EersWABpkJxslKGZR5ABeAD5kAnZgGzHkObEqNY2t3a8rAQTlwp3OWEu0BumzMzDmaRMWxG9yeLzeHy+5HUbW0CRQKSgCxyDmWRR+6xh4AB+w4h2OZwuVxuyMez1e73GQm+yF+lJ2e28wPyYMZUOQf1hXSgCIK4BZqPZHwJrSU5gCcM493FpFUyG6n04IzEAQAdL8DkcwbdWTqjeRTeahc4rfKyKo7eRVcgqpzCXdtbr9QS7VUzesLfTxa7bWJQ46QRAXSi3R6YPUQAgwMA1uDbEoMmA7EN2PwAFYa+Z8YNknllsNDCNJm3u7kl0v1lhO0FnCXgCsygu8KUDpHJmNYoA) doesn't help the compiler resolve the overload.

[Defining a base signature for the overloaded method](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIAs4gOYQKIBuE4yA3gFDLJgCeADhAFzIDkMA9uy8gD6sBGcKCwDclZBAC2dWs36cANhCxiAvuXKhIsRCgBinDFlyFiYCQA9IIACYBnNJhz4iJClVoNmbTqPEd2AEkQDmY7MChQbDUNLWh4JGQAISEjZ1MSCCtie0djFzMycU8mASE-KkEoYNDkEABXSX5oGM1weN1kuDsUd2QEJSEACgBKZhSesSp2ECGIVzAAFXpSnwR6uxYAGmQnGyUoZlHkAF4APmQCdmAbMa7J8Rm5heWvVgQnXG3drH3oI-mZmYaRMCxGpwuVxudwmEFacR0iQM7Es1lysKK01mgPAr1WHHWmx2ewOR3B50u11uzGRU2QTxxSxW3g++W+JP+yGeQOQyJBBXA5MhVLutPI6ja2gSKBSUFROQcGL6DJezNYBI27N+pK5Qsp0PGQjpKrMeJZnwgWtsOu54ENUH5GTAeqh1K6UFaSnMARpnFOXNIqmQ3V5nBGYgCADonmtNTtjhTA+HyFGY6znN8Exck2JyF7kFV7f6hoHg4qhMmqtHZrGibqIWRVJWhNWhix01943qcxoYPUQAgwMAZj9rQKwHYhux+AArZgAegAVPcUIv56GUXxZeDlbPW7XM92m8a92mLYeGz3VEA) only helps if using `Base` and not a union of subtypes, which isn't acceptable if one wants to only accept a subset of subtypes.
